gunpowder/nodes/wkw_source.py

```diff
@@ -0,0 +1,188 @@
+                raise RuntimeError("%s not in %s" % (ds_name, self.filename))
+
+            spec = self.__read_spec(array_key, data_file, ds_name)
+
+            self.provides(array_key, spec)
+
+    def provide(self, request):
+
+        timing = Timing(self)
+        timing.start()
+
+        batch = Batch()
+
+        data_file = self._open_file(self.filename)
+        for (array_key, request_spec) in request.array_specs.items():
+            logger.debug("Reading %s in %s...", array_key, request_spec.roi)
+
+            voxel_size = self.spec[array_key].voxel_size
+
+            # scale request roi to voxel units
+            dataset_roi = request_spec.roi / voxel_size
+
+            # shift request roi into dataset
+            # Comment(erjel): Currently I do not see an application for this.
+            # dataset_roi = dataset_roi \
+            #    - self.spec[array_key].roi.get_offset() / voxel_size
+
+
+            # create array spec
+            array_spec = self.spec[array_key].copy()
+            array_spec.roi = request_spec.roi
+
+
+            # add array to batch
+            batch.arrays[array_key] = Array(
+                self.__read(
+                    data_file,
+                    self.datasets[array_key],
+                    self.mag_specs[array_key],
+                    dataset_roi
+                ),
+                array_spec)
+
+        logger.debug("done")
+
+        timing.stop()
+        batch.profiling_stats.add(timing)
+
+        return batch
+
+    def _get_voxel_size(self, layer, mag_spec):
+        mag = Mag(mag_spec).to_np()
+        dset_vox_size = np.array(layer.dataset.scale)
+        return Coordinate(mag*dset_vox_size)
+
+    def _get_offset(self, layer):
+        return Coordinate(layer._properties.bounding_box.topleft)
+
+    def __read_spec(self, array_key, data_file, ds_name):
+
+        layer = data_file.get_layer(ds_name)
+
+        if array_key in self.array_specs:
+            spec = self.array_specs[array_key].copy()
+        else:
+            spec = ArraySpec()
+
+        if spec.voxel_size is None:
+            voxel_size = self._get_voxel_size(layer, self.mag_specs[array_key])
+            if voxel_size is None:
+                voxel_size = Coordinate((1,)*len(layer.dataset.shape))
+                logger.warning(
+                    "WARNING: File %s does not contain resolution information "
+                    "for %s (dataset %s), voxel size has been set to %s. This "
+                    "might not be what you want.",
+                    self.filename, array_key, ds_name, spec.voxel_size
+                )
+            spec.voxel_size = voxel_size
+
+        self.ndims = len(spec.voxel_size)
+
+        if spec.roi is None:
+            offset = self._get_offset(layer)
+            offset *= spec.voxel_size
+            if offset is None:
+                offset = Coordinate((0,)*self.ndims)
+
+            shape = Coordinate(layer._properties.bounding_box.size)
```

Comment:
I would guess that the shape is provided in voxels at the highest resolution. In the next line you have `spec.roi = Roi(offset, shape*spec.voxel_size)` but it looks like the voxel_size accounts for the magnification. I think using `Coordinate(layer.dataset.scale)` instead of `spec.voxel_size` would fix it.
If downsampling by your magnification cannot be done evenly then the provided roi at some resolutions might include fractions of a voxel on the upper or lower end. The `Roi` has a helper function `snap_to_grid` which takes a voxel size and will crop off the voxel parts.